dalm/datasets/reading_comprehension_generation/synthetic_based.py

```diff
@@ -0,0 +1,103 @@
+    return [
+        {
+            "role": "system",
+            "content": "You are a helpful and meticulous instruction following question and answer making chatbot. Please refrain from acknowledgments, additions or niceties of any sort",
+        },
+        {"role": "user", "content": prompt},
+    ]
+
+
+def generate_synthetic_data(model_pipeline, text, generation_params):
+    prompt = gen_prompt(text)
+    prompt = model_pipeline.tokenizer.apply_chat_template(prompt, tokenize=False, add_generation_prompt=True)
+    outputs = model_pipeline(prompt, **generation_params)
+
+    return outputs[0]["generated_text"]
+
+
+def generate_synthetic_dataset(
+    model_name,
+    input_directory,
+    state_file,
+    chunk=False,
+    context_length=2048,
+    generation_params={
+        "max_new_tokens": 600,
+        "do_sample": True,
+        "temperature": 0.7,
+        "top_k": 5,
+        "top_p": 0.95,
+        "return_full_text": False,
+    },
+):
+    model_pipeline = pipeline("text-generation", model=model_name, torch_dtype=torch.bfloat16, device_map="auto")
+
+    input_files = list_dir(input_directory)
+
+    if chunk:
+        tokenizer = model_pipeline.tokenizer
+        tokens = tokenizer.apply_chat_template(gen_prompt(""), tokenize=False, add_generation_prompt=True)
+        CONSTANT = len(tokenizer(tokens)["input_ids"])
+        k = context_length - CONSTANT
+
+    if os.path.exists(state_file):
+        with open(state_file, "rb") as f:
+            state = pickle.load(f)
+    else:
+        state = {"processed_files": []}
+        pickle.dump(state, open(state_file, "wb"))
+
+    for file, text in input_files:
+        if file in state["processed_files"]:
+            continue
+        else:
+            if chunk:
+                for chunk_ in text_chunker(text, tokenizer, k):
+                    gen_text = generate_synthetic_data(model_pipeline, chunk_, generation_params)
+                    yield gen_text, chunk_
+            else:
+                gen_text = generate_synthetic_data(model_pipeline, text, generation_params)
+                yield gen_text, text
+
+            state["processed_files"].append(file)
+            pickle.dump(state, open(state_file, "wb"))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", type=str, required=True)
+    parser.add_argument("--input_directory", type=str, required=True)
+    parser.add_argument("--output_directory", type=str, required=True)
+    parser.add_argument("--state_file", type=str, required=True)
```

Comment:
Any chance `state_file` could be made optional?  Meaning if the user didn't pass anything, the code would just create a state file as a temp file somewhere?

State file = something to track processing state?

Reply:
Also, is a separate processing state file even needed?  If the file has been processed, won't it be present in the output directory?  (and likewise, if not processed, not present in the output directory)

Reply:
I was about to remove the state file. 
I think because the function has no control over the naming of the output file, keeping a record of the files already processed is the most reliant way of keeping record

Also made a minor change giving user the ability to switch off behaviour

Reply:
:thinking: that said, it would be cleaner to give the caller of the function the reins to state keeping of this nature

Reply:
> I think because the function has no control over the naming of the output file, keeping a record of the files already processed is the most reliant way of keeping record

Yeah for this to work:

> If the file has been processed, won't it be present in the output directory? (and likewise, if not processed, not present in the output directory)

The output file would need to have the same name as the input file, or use a "content addressable" scheme of some sort (hash of content).  Currently it uses an index in the filename.

If there's a 1:1 mapping between input and output file, then using the same filename in both input and output directories should make it easy to track the processing state without needing an extra file.

Reply:
> give the caller of the function the reins to state keeping of this nature

I think only a small percentage of users would want any control here.  So it would be better to have the default behavior "just work" while allowing power users to override it if needed.

Reply:
At the time you reported this, csv file were considered a valid input, it's very likely the processing was correct and the one file didn't make it through because the q&a parser didn't parse anything out from  output of the llm

Reply:
@metric-space Ok I will try this again and see if I observe the same issue